### PR TITLE
Add support for autodns

### LINKED ISSUE #501: AutoDNS API Support?
Hi Folks,
I appreciate your efforts and will appreciate much more if lego supports AutoDNS API. There is already a library which written in PHP here:
https://github.com/tillkahlbrock/autodns-api
And the AutoDNS XML API documentation:
https://help.internetx.com/display/API/AutoDNS+XML-API

### PR DESCRIPTION
**Pull request by kolaente**
This pr adds support for autodns based on https://help.internetx.com/display/APIJSONEN

Closes #501

### Checklist

* [x] add a description to your PR
* [x] be able to maintain this provider
* [x] have a homogeneous design with the other providers
* [x] add tests (units)
* [x] add tests ("live") (i guess? Please correct me if I'm wrong)
* [x] add the provider to the readme.md
* [x] add a provider descriptor
* [x] generate CLI help and documentation
* [x] pass the linter (golangci-lint must be installed)
* [x] do `go mod tidy`
* [x] be able to do: (and put the output of this command to a comment in your PR)
```
rm -rf .lego

./lego -m your@email.com --dns YOUR_PROVIDER_NAME -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```
Note the wildcard domain is important.

### CODE REVIEW

File: cmd/zz_gen_cmd_dnshelp.go [hunk 1]
```diff
@@ -142,6 +143,22 @@ func displayDNSHelp(name string) error {
...
+		ew.writeln()
+
+		ew.writeln(`Credentials:`)
+		ew.writeln(`	- "AUTODNS_API_PASSWORD":	User Password`)
+		ew.writeln(`	- "AUTODNS_API_USER":	User ID`)
+		ew.writeln(`	- "AUTODNS_ENDPOINT":	API endpoint URL`)
+		ew.writeln()
+
+		ew.writeln()
+		ew.writeln(`More information: https://go-acme.github.io/lego/dns/autodns`)
```
Comment (amenk) [inline cmd/zz_gen_cmd_dnshelp.go]: does this page need to be created?

File: providers/dns/autodns/autodns.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	config := NewDefaultConfig()
+	config.Username = values[envAPIUser]
+	config.Password = values[envAPIPassword]
+	config.Endpoint = endpoint
+
+	provider := &DNSProvider{config: config}
+
+	// Because autodns needs the nameservers for each request, we query them all here and put them
+	// in our state to avoid making a lot of requests later.
+	// FIXME: This should become obsolete once I figure out how the _stream endpoint works.
```
Comment (amenk) [inline providers/dns/autodns/autodns.go]: did you ask AutoDNS developers?

Comment (kolaente) [reply providers/dns/autodns/autodns.go]: Yes, I sent them a mail.

File: cmd/zz_gen_cmd_dnshelp.go (hunk 1 above)
Comment (kolaente) [reply cmd/zz_gen_cmd_dnshelp.go]: I think so, yes. It is autogenerated. Maybe one of the LEGO maintainers can verify.

File: providers/dns/autodns/autodns.go (hunk 2 above)
Comment (kolaente) [reply providers/dns/autodns/autodns.go]: I used it like in their documentation, but it did not work, so we'll just wait what they have to say.
The challenge works though if you check out 1b8f6763d55523bb27e8329055b39b708dcc6cc2

**Comment by ldez** [comment]
Hello, in order for a PR adding a DNS provider to be accepted, you have to:

- [x] add a description to your PR
- [x] be able to maintain this provider
- [x] have a homogeneous design with the other providers
- [x] add tests (units)
```shell
make test
```
- [x] add tests ("live")
https://github.com/go-acme/lego/blob/c938de68f262c057542d44bcb67ef961cefcc60e/providers/dns/cloudflare/cloudflare_test.go#L125-L151

```shell
make test
```
- [x] add the provider to the [readme.md](https://github.com/go-acme/lego#dns-providers)
- [x] add a [provider descriptor](https://github.com/go-acme/lego/blob/master/providers/dns/cloudflare/cloudflare.toml)
- [x] generate CLI help and documentation
```bash
make generate-dns
```
- [x] be able to do: _(and put the output of this command to a comment in your PR)_
```bash
rm -rf .lego

./lego -m your@email.com --dns YOUR_PROVIDER_NAME -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```
Note the wildcard domain is important.
- [x] pass the linter ([golangci-lint](https://github.com/golangci/golangci-lint#install) must be installed):
```shell
make checks
```
- [x] do `go mod tidy`

File: cmd/zz_gen_cmd_dnshelp.go (hunk 1 above)
Comment (ldez) [reply cmd/zz_gen_cmd_dnshelp.go]: this page automatically created

**Comment by kolaente** [comment]
@ldez I updated the original description of the pr with your checklist.

**Comment by kolaente** [comment]
I also go a response about the `_stream` api endpoint from the autodns guys, they could reproduce it and are investigating why it did not work - they'll get back to me later today.

File: providers/dns/autodns/autodns.toml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+Name = "Autodns"
+Description = ''''''
+URL = "https://www.internetx.com/domains/autodns/"
+Code = "autodns"
+Since = "v3.1.0"
+
+Example = ''''''
```
Comment (ldez) [inline providers/dns/autodns/autodns.toml]: could you add an example.

File: providers/dns/autodns/autodns.toml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+[Configuration]
+  [Configuration.Credentials]
+    AUTODNS_API_USER = "Username"
+    AUTODNS_API_PASSWORD = "User Password"
+    AUTODNS_ENDPOINT = "API endpoint URL"
+    AUTODNS_CONTEXT = "API context (4 for production, 1 for testing. Defaults to 4)"
+    AUTODNS_TTL = "The TTL of the TXT record used for the DNS challenge"
```
Comment (ldez) [inline providers/dns/autodns/autodns.toml]: could you follow the same pattern as other providers:
https://github.com/go-acme/lego/blob/a5a29187fea9f555ea43fae16ca65066c452d330/providers/dns/auroradns/auroradns.toml#L9-L17

File: providers/dns/autodns/autodns.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	config := NewDefaultConfig()
+	config.Username = values[envAPIUser]
+	config.Password = values[envAPIPassword]
+	config.Endpoint = endpoint
+	config.Context = endpointContext
+	config.TTL = ttl
+
+	provider := &DNSProvider{config: config}
+
+	// Because autodns needs the nameservers for each request, we query them all here and put them
```
Comment (ldez) [inline providers/dns/autodns/autodns.go]: could you follow the same pattern as other provider: `NewDNSProvider`, `NewDNSProviderConfig`

File: providers/dns/autodns/autodns.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	HTTPClient *http.Client
+}
+
+func NewDefaultConfig() *Config {
+	endpoint, _ := url.Parse(defaultEndpoint)
+
+	return &Config{
+		Endpoint:   endpoint,
+		Context:    defaultEndpointContext,
+		HTTPClient: &http.Client{},
```
Comment (ldez) [inline providers/dns/autodns/autodns.go]: could you follow the same pattern as other provider:
https://github.com/go-acme/lego/blob/a5a29187fea9f555ea43fae16ca65066c452d330/providers/dns/cloudflare/cloudflare.go#L32-L41

**Comment by kolaente** [comment]
@ldez When implementing `Timeout()`, should I make it configurable?

File: providers/dns/autodns/autodns.toml (hunk 3 above)
Comment (kolaente) [reply providers/dns/autodns/autodns.toml]: Done.

File: providers/dns/autodns/autodns.toml (hunk 4 above)
Comment (kolaente) [reply providers/dns/autodns/autodns.toml]: Done.

File: providers/dns/autodns/autodns.go (hunk 6 above)
Comment (kolaente) [reply providers/dns/autodns/autodns.go]: Done.

File: providers/dns/autodns/autodns.go (hunk 5 above)
Comment (kolaente) [reply providers/dns/autodns/autodns.go]: Done.

**Comment by kolaente** [comment]
@ldez I've added some tests, I'll add some more for `NewDNSProviderConfig` once I know how to proceed with the `_stream` endpoint.

**Comment by kolaente** [comment]
They got back to me, looks like it was a bug in their system. They will roll out a fix at the beginning of the next week.

**Comment by kolaente** [comment]
@ldez It looks like the CI failure was caused by Go 1.13 behaving different, could you check?

**Comment by kolaente** [comment]
CI failure is reproducable with go 1.13 on my local env

**Comment by kolaente** [comment]
Looks like golangci is well aware of the issue: https://github.com/golangci/golangci-lint/issues/652

**Comment by ldez** [comment]
see #962

**Comment by ldez** [comment]
any news?

Note: #962 is merged

**Comment by kolaente** [comment]
@ldez Yes, autdns fixed their api endpoint, I'll take a look on it today, I was on vacation until yesterday.

**Comment by kolaente** [comment]
Looks like there is still an issue with their api, all new txt records I create get inserted with `*` as name. The challenge itself works, but cleanup does not. I've sent them a mail about this.

**Comment by ldez** [comment]
@kolaente any news?

**Comment by kolaente** [comment]
@ldez Unfortunatly not, I've sent them a mail today asking about a status update. I'll keep you updated here.

**Comment by kolaente** [comment]
They will put a fix live this week.

**Comment by kolaente** [comment]
They've put the fix live! There still was a small bug in my implementation, which I fixed. Challenge and cleanup works now.

**Comment by kolaente** [comment]
Wildcard cert test:

```
$ rm -fr .lego && AUTODNS_API_USER=[redacted] AUTODNS_API_PASSWORD=[redacted] ./dist/lego --server=https://acme-staging-v02.api.letsencrypt.org/directory -d '*.example.com' -m mail@example.com --dns autodns -a run
2019/10/25 15:28:13 No key found for account mail@example.com. Generating a P384 key.
2019/10/25 15:28:13 Saved key to ~/go/src/github.com/go-acme/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/mail@example.com/keys/mail@example.com.key
2019/10/25 15:28:13 [INFO] acme: Registering account for mail@example.com
!!!! HEADS UP !!!!

		Your account credentials have been saved in your Let's Encrypt
		configuration directory at "~/go/src/github.com/go-acme/lego/.lego/accounts".
		You should make a secure backup	of this folder now. This
		configuration directory will also contain certificates and
		private keys obtained from Let's Encrypt so making regular
		backups of this folder is ideal.2019/10/25 15:28:14 [INFO] [*.example.com] acme: Obtaining bundled SAN certificate
2019/10/25 15:28:14 [INFO] [*.example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/16726270
2019/10/25 15:28:14 [INFO] [*.example.com] acme: use dns-01 solver
2019/10/25 15:28:14 [INFO] [*.example.com] acme: Preparing to solve DNS-01
2019/10/25 15:28:15 [INFO] [*.example.com] acme: Trying to solve DNS-01
2019/10/25 15:28:15 [INFO] [*.example.com] acme: Checking DNS record propagation using [127.0.0.53:53]
2019/10/25 15:28:15 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2019/10/25 15:28:21 [INFO] [*.example.com] The server validated our request
2019/10/25 15:28:21 [INFO] [*.example.com] acme: Cleaning DNS-01 challenge
2019/10/25 15:28:24 [INFO] [*.example.com] acme: Validations succeeded; requesting certificates
2019/10/25 15:28:25 [INFO] [*.example.com] Server responded with a certificate.
```

**Comment by kolaente** [comment]
CI passes now.

**Comment by kolaente** [comment]
@ldez What else needs to be done to get this PR merged?

**Comment by ldez** [comment]
Could you run:
```
./lego -m your@email.com --dns autodns -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```

Note, there are 2 domains ` -d *.example.com -d example.com`

**Comment by kolaente** [comment]
@ldez I've updated the comment.

**Comment by ldez** [comment]
I'm annoyed because there is a bug with GitHub (I contacted them on this topic) that prevents me from editing your PR because your fork is in an organization.

So there is 3 solutions:
- you give the rights on your repository (I do not feel very comfortable with this because it's a bit extreme as a solution)
- I open a PR on your fork to edit your code. (a PR on PR, it's a bit weird)
- you open a new PR by using a fork in your account. (I dislike this solution but it's a solution)

What do you prefer?

**Comment by kolaente** [comment]
@ldez what do you need the rights for?

**Comment by ldez** [comment]
I need the rights to apply some review fixes.

I opened a PR https://github.com/iMi-digital/lego/pull/1

**Comment by kolaente** [comment]
@ldez Merged.

**Comment by ldez** [comment]
could you solve this:
```
golangci-lint run
providers/dns/autodns/client.go:16:2: `demoEndpoint` is unused (varcheck)
	demoEndpoint    = "https://api.demo.autodns.com/v1/"
	^
Makefile:39: recipe for target 'checks' failed
```

**Comment by kolaente** [comment]
@ldez Done.

**Comment by kolaente** [comment]
@ldez How do we proceed to get this in traefik? Should I send a pr after you released a new version of lego?

**Comment by ldez** [comment]
yes you have to wait for the next lego release.

**Comment by rmoriz** [comment]
This implementation seems not work with LEGO_EXPERIMENTAL_CNAME_SUPPORT

**Comment by kolaente** [comment]
@rmoriz Could you open an issue?

**Pull request merged**